File: android/app/src/main/java/com/example/asteroidzen/WidgetFeedPlugin.java
package com.example.asteroidzen;

import android.util.Base64;

import com.example.asteroidzen.widget.BaseWidgetProvider;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;

/**
 * In-app Capacitor plugin: receives pre-rendered widget frames + manifest
 * from the web layer and stores them where the widget provider (same
 * process/uid) can read them: filesDir/widgets/.
 */
@CapacitorPlugin(name = "WidgetFeed")
public class WidgetFeedPlugin extends Plugin {

  public static final String DIR = "widgets";
  public static final String MANIFEST = "manifest.json";

  private File widgetsDir() {
    File dir = new File(getContext().getFilesDir(), DIR);
    if (!dir.exists()) dir.mkdirs();
    return dir;
  }

  @PluginMethod
  public void write(PluginCall call) {
    try {
      File root = widgetsDir();
      JSArray files = call.getArray("files");
      if (files != null) {
        for (int i = 0; i < files.length(); i++) {
          JSONObject f = files.getJSONObject(i);
          String rel = f.getString("path");
          if (rel.contains("..")) continue; // stay inside the feed dir
          byte[] bytes = Base64.decode(f.getString("dataB64"), Base64.DEFAULT);
          File out = new File(root, rel);
          File parent = out.getParentFile();
          if (parent != null && !parent.exists()) parent.mkdirs();
          try (FileOutputStream fos = new FileOutputStream(out)) {
            fos.write(bytes);
          }
        }
      }
      String manifestJson = call.getString("manifestJson");
      if (manifestJson != null) {
        // manifest written last + atomically: widgets never see a half feed
        File tmp = new File(root, MANIFEST + ".tmp");
        try (FileOutputStream fos = new FileOutputStream(tmp)) {
          fos.write(manifestJson.getBytes(StandardCharsets.UTF_8));
        }
        tmp.renameTo(new File(root, MANIFEST));
        pruneStaleBaseDirs(root, manifestJson);
      }
      call.resolve();
    } catch (Exception e) {
      call.reject("widget feed write failed", e);
    }
  }

  @PluginMethod
  public void reloadWidgets(PluginCall call) {
    BaseWidgetProvider.updateAll(getContext());
    BaseWidgetProvider.ensureWorker(getContext());
    call.resolve();
  }

  private void pruneStaleBaseDirs(File root, String manifestJson) throws IOException {
    Set<String> keep = new HashSet<>();
    try {
      JSONObject manifest = new JSONObject(manifestJson);
      org.json.JSONArray bases = manifest.getJSONArray("bases");
      for (int i = 0; i < bases.length(); i++) {
        keep.add("b_" + bases.getJSONObject(i).getString("panelKey").replace(",", "_"));
      }
    } catch (Exception ignored) {
      return;
    }
    File[] children = root.listFiles();
    if (children == null) return;
    for (File c : children) {
      if (c.isDirectory() && c.getName().startsWith("b_") && !keep.contains(c.getName())) {
        File[] frames = c.listFiles();
        if (frames != null) for (File f : frames) f.delete();
        c.delete();
      }
    }
  }
}

File: ios/App/App/WidgetFeedPlugin.swift
import Capacitor
import Foundation
import WidgetKit

/// In-app Capacitor plugin: receives pre-rendered widget frames + manifest
/// from the web layer and stores them in the App Group container where the
/// BaseWidget extension reads them.
///
/// NOTE: the group id embeds the (placeholder) bundle id — keep in sync with
/// BaseWidget/WidgetManifest.swift and rename both before store submission.
@objc(WidgetFeedPlugin)
public class WidgetFeedPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetFeedPlugin"
    public let jsName = "WidgetFeed"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "write", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reloadWidgets", returnType: CAPPluginReturnPromise),
    ]

    static let appGroupId = "group.com.example.asteroidzen.widgets"

    private func feedDir() -> URL {
        let container =
            FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: Self.appGroupId)
            // unsigned dev builds have no App Group — fall back so the app still runs
            ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let dir = container.appendingPathComponent("widgets", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    @objc func write(_ call: CAPPluginCall) {
        let root = feedDir()
        do {
            for entry in call.getArray("files", JSObject.self) ?? [] {
                guard
                    let rel = entry["path"] as? String, !rel.contains(".."),
                    let b64 = entry["dataB64"] as? String,
                    let bytes = Data(base64Encoded: b64)
                else { continue }
                let out = root.appendingPathComponent(rel)
                try FileManager.default.createDirectory(
                    at: out.deletingLastPathComponent(), withIntermediateDirectories: true)
                try bytes.write(to: out)
            }
            if let manifestJson = call.getString("manifestJson") {
                // manifest written last + atomically: the widget never sees a half feed
                let manifestURL = root.appendingPathComponent("manifest.json")
                try Data(manifestJson.utf8).write(to: manifestURL, options: .atomic)
                pruneStaleBaseDirs(root: root, manifestJson: manifestJson)
            }
            call.resolve()
        } catch {
            call.reject("widget feed write failed: \(error.localizedDescription)")
        }
    }

    @objc func reloadWidgets(_ call: CAPPluginCall) {
        WidgetCenter.shared.reloadAllTimelines()
        call.resolve()
    }

    private func pruneStaleBaseDirs(root: URL, manifestJson: String) {
        guard
            let data = manifestJson.data(using: .utf8),
            let manifest = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let bases = manifest["bases"] as? [[String: Any]]
        else { return }
        let keep = Set(
            bases.compactMap { ($0["panelKey"] as? String).map { "b_" + $0.replacingOccurrences(of: ",", with: "_") } }
        )
        let children =
            (try? FileManager.default.contentsOfDirectory(at: root, includingPropertiesForKeys: [.isDirectoryKey]))
            ?? []
        for child in children where child.lastPathComponent.hasPrefix("b_") {
            if !keep.contains(child.lastPathComponent) {
                try? FileManager.default.removeItem(at: child)
            }
        }
    }
}
